execution/gethexec/express_lane_service.go

```diff
@@ -0,0 +1,410 @@
+						log.Warn("New express lane controller did not match previous controller",
+							"round", round,
+							"previous", setExpressLaneIterator.Event.PreviousExpressLaneController,
+							"new", setExpressLaneIterator.Event.NewExpressLaneController)
+						continue
+					}
+					newController := setExpressLaneIterator.Event.NewExpressLaneController
+					es.roundControl.Add(round, &expressLaneControl{
+						controller: newController,
+						sequence:   0,
+					})
+				}
+				fromBlock = toBlock
+			}
+		}
+	})
+}
+
+func (es *expressLaneService) currentRoundHasController() bool {
+	currRound := timeboost.CurrentRound(es.initialTimestamp, es.roundDuration)
+	control, ok := es.roundControl.Get(currRound)
+	if !ok {
+		return false
+	}
+	return control.controller != (common.Address{})
+}
+
+func (es *expressLaneService) isWithinAuctionCloseWindow(arrivalTime time.Time) bool {
+	// Calculate the next round start time
+	elapsedTime := arrivalTime.Sub(es.initialTimestamp)
+	elapsedRounds := elapsedTime / es.roundDuration
+	nextRoundStart := es.initialTimestamp.Add((elapsedRounds + 1) * es.roundDuration)
+	// Calculate the time to the next round
+	timeToNextRound := nextRoundStart.Sub(arrivalTime)
+	// Check if the arrival timestamp is within AUCTION_CLOSING_DURATION of TIME_TO_NEXT_ROUND
+	return timeToNextRound <= es.auctionClosing
+}
+
+// Sequence express lane submission skips validation of the express lane message itself,
+// as the core validator logic is handled in `validateExpressLaneTx“
+func (es *expressLaneService) sequenceExpressLaneSubmission(
+	ctx context.Context,
+	msg *timeboost.ExpressLaneSubmission,
+) error {
+	// no service lock needed since roundControl is thread-safe
+	control, ok := es.roundControl.Get(msg.Round)
+	if !ok {
+		return timeboost.ErrNoOnchainController
+	}
+
+	es.Lock()
+	defer es.Unlock()
+	// Check if the submission nonce is too low.
+	if msg.SequenceNumber < control.sequence {
+		return timeboost.ErrSequenceNumberTooLow
+	}
+	// Check if a duplicate submission exists already, and reject if so.
+	if _, exists := es.messagesBySequenceNumber[msg.SequenceNumber]; exists {
+		return timeboost.ErrDuplicateSequenceNumber
+	}
+	// Log an informational warning if the message's sequence number is in the future.
+	if msg.SequenceNumber > control.sequence {
+		log.Info("Received express lane submission with future sequence number", "SequenceNumber", msg.SequenceNumber)
+	}
+	// Put into the sequence number map.
+	es.messagesBySequenceNumber[msg.SequenceNumber] = msg
+
+	for {
+		// Get the next message in the sequence.
+		nextMsg, exists := es.messagesBySequenceNumber[control.sequence]
+		if !exists {
+			break
+		}
+		if err := es.transactionPublisher.PublishTimeboostedTransaction(
+			ctx,
+			nextMsg.Transaction,
+			msg.Options,
+		); err != nil {
+			// If the tx failed, clear it from the sequence map.
+			delete(es.messagesBySequenceNumber, msg.SequenceNumber)
+			return err
+		}
```

Comment:
Are we doing anything with messagesBySequenceNumber? any reason to keep messages in the map if err == nil?